### ISSUE TITLE
Post-epoch hooks?

### ISSUE THREAD

**Issue by zabzug-pfpt** [KevinMusgrave/pytorch-metric-learning#2]
Great package! Do you have any plans to add post-epoch hooks that would allow the user to do things like save the model state_dict or run on a validation set?

**Comment by KevinMusgrave** [comment]
Thanks for the suggestion! Here's how I'm thinking of implementing it within the BaseTrainer class.
```python
class BaseTrainer:
    def __init__(self, ..., end_of_epoch_hook=None):
        self.end_of_epoch_hook = (lambda x: None) if end_of_epoch_hook is None else end_of_epoch_hook 

    def train(self, start_epoch=1, num_epochs=0):
        self.set_to_train()
        for self.epoch in range(start_epoch, start_epoch+num_epochs):
            logging.info("TRAINING EPOCH %d" % self.epoch)
            pbar = tqdm.tqdm(range(self.iterations_per_epoch))
            for self.iteration in pbar:
                self.forward_and_backward()
                pbar.set_description("total_loss=%.5f" % self.losses["total_loss"])
            self.step_lr_schedulers()
            self.update_records(end_of_epoch=True)
            self.end_of_epoch_hook(self)
```

Then to create a hook:
```python
def your_hook(self):
    # Access any of the trainer's attributes. For example:
    trunk = self.models["trunk"]
    # Save model etc
    ....
```
(I'm calling the input "self" even though this is defined outside of the class.)

Now that I think of it, I could probably convert all the "record_keeper" stuff to this format as well, i.e. let the user define hooks for updating logs their own way. 

Let me know what you think.

**Comment by zabzug-pfpt** [comment]
That looks pretty good to me!

**Comment by KevinMusgrave** [comment]
Latest version has this feature. See [this script](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/examples/example_MetricLossOnly.py#L74) for an example.

**Comment by KevinMusgrave** [comment]
@zabzug-pfpt In the latest version (v0.9.68), if your end_of_epoch_hook returns False, then training will stop early. (It has to return False, not None or 0 etc.)
